File: frontend/src/types/llm.ts
export type RoutingHint = 'fast' | 'reasoning' | 'large_context' | 'low_budget' | 'offline';

export const ROUTING_HINTS: { value: RoutingHint; label: string }[] = [
  { value: 'fast', label: 'Fast' },
  { value: 'reasoning', label: 'Reasoning' },
  { value: 'large_context', label: 'Large context' },
  { value: 'low_budget', label: 'Low budget' },
  { value: 'offline', label: 'Offline (local)' },
];

export interface ModelSpec {
  provider: string;
  model: string;
  context_window: number;
  price_per_1m_input: number;
  price_per_1m_output: number;
  supports_streaming: boolean;
  supports_json_mode: boolean;
  supports_vision: boolean;
  supports_function_calling: boolean;
  supports_reasoning: boolean;
  is_fast: boolean;
  is_reasoning: boolean;
}

export interface ProviderHealth {
  provider: string;
  configured: boolean;
  healthy: boolean;
}

export type LLMRequestStatus = 'pending' | 'completed' | 'failed';

export interface AttemptRecord {
  provider: string;
  model: string;
  error: string | null;
}

export interface LLMRequest {
  id: string;
  prompt_id: string;
  provider: string;
  model: string;
  routing_hint: string | null;
  input_text: string;
  output_text: string | null;
  input_tokens: number;
  output_tokens: number;
  cost_usd: number | null;
  latency_ms: number | null;
  stream: boolean;
  json_mode: boolean;
  attempted_providers: AttemptRecord[] | null;
  status: LLMRequestStatus;
  status_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateCompletionRequest {
  provider?: string | null;
  model?: string | null;
  routing_hint?: RoutingHint | null;
  json_mode?: boolean;
}

File: backend/app/schemas/llm.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, model_validator

from app.models.llm_request import LLMRequestStatus


class ModelSpecRead(BaseModel):
    provider: str
    model: str
    context_window: int
    price_per_1m_input: float
    price_per_1m_output: float
    supports_streaming: bool
    supports_json_mode: bool
    supports_vision: bool
    supports_function_calling: bool
    supports_reasoning: bool
    is_fast: bool
    is_reasoning: bool


class ProviderHealth(BaseModel):
    provider: str
    configured: bool
    healthy: bool


class LLMRequestRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    prompt_id: uuid.UUID
    provider: str
    model: str
    routing_hint: str | None
    input_text: str
    output_text: str | None
    input_tokens: int
    output_tokens: int
    cost_usd: float | None
    latency_ms: int | None
    stream: bool
    json_mode: bool
    attempted_providers: list[dict] | None
    status: LLMRequestStatus
    status_message: str | None
    created_at: datetime
    updated_at: datetime


class CreateCompletionRequest(BaseModel):
    provider: str | None = None
    model: str | None = None
    routing_hint: str | None = None
    json_mode: bool = False

    @model_validator(mode="after")
    def _provider_and_model_together(self) -> "CreateCompletionRequest":
        if bool(self.provider) != bool(self.model):
            raise ValueError(
                "provider and model must be given together, or both omitted to use "
                "routing_hint (or the default model) instead."
            )
        return self
